### PR TITLE
AggregationISM set for rc kathy

### LINKED ISSUE #2224: Use 1/2 aggregation of merkle proof or message ID ISM on testnet rc kathy

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- Using a separate set of validators for RC deployments and tests
- Defining a recursive AggregationISM config which aggregates over two RoutingISMs, one for a set of MerkleRootMultisigISMs and one for a set of MessageIdMultisigISMs for each origin domain.

### Drive-by changes

- Dployed MerkleRootMultisigISMFactory and MessageIdMultisigISMFactory for testnet3 in hyperlane context

### Related issues

- Fixes #2224

### Backward compatibility

 Backward compatible?
 Yes

Any infrastructure implications? 
HyperlaneRouterChecker will fail silently if you provide an IsmConfig - #2534

### Testing

Manual
Testnet

### CODE REVIEW

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 1]
```diff
@@ -182,12 +184,26 @@ export abstract class HyperlaneDeployer<
...
+            local,
+            connectionClient.setInterchainSecurityModule(
+              config.interchainSecurityModule,
+            ),
+          );
+        } else if (this.ismFactory) {
+          // deploy matching module
+          const ism = await this.ismFactory.deploy(
+            local,
             config.interchainSecurityModule,
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/HyperlaneDeployer.ts:198]: this could still be `undefined` here IIUC

**Review by asaj — COMMENTED** [review]
Took a quick look

File: typescript/infra/scripts/helloworld/deploy.ts [hunk 2]
```diff
@@ -17,12 +17,19 @@ import {
...
+  console.log(coreConfig);
   // Always deploy from the hyperlane deployer
   const multiProvider = await coreConfig.getMultiProvider(
     Contexts.Hyperlane,
     Role.Deployer,
   );
   const configMap = await getRouterConfig(environment, multiProvider, true);
+  console.log('ConfigMap: ', configMap);
+
+  // for (const routingIsms of Object.entries( aggregationIsm().modules)) {
```
Comment (jmrossy) [inline typescript/infra/scripts/helloworld/deploy.ts:30]: dead code?

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 3]
```diff
@@ -45,6 +49,7 @@ export abstract class HyperlaneDeployer<
   constructor(
     protected readonly multiProvider: MultiProvider,
     protected readonly factories: Factories,
+    protected readonly ismFactory?: HyperlaneIsmFactory,
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply typescript/sdk/src/deploy/HyperlaneDeployer.ts:52]: I don't think so. Factores is contractFactory and HyperlaneIsmFactory is ismFactoryFactories

File: typescript/infra/scripts/helloworld/deploy.ts (hunk 2 above)
Comment (aroralanuk) [reply typescript/infra/scripts/helloworld/deploy.ts:30]: good catch

File: typescript/sdk/src/router/HyperlaneRouterChecker.ts [hunk 4]
```diff
@@ -39,8 +39,12 @@ export class HyperlaneRouterChecker<
       violationType: ConnectionClientViolationType,
     ) => {
       const actual = await router[property]();
+      // TODO: check for IsmConfig
       const expected =
-        this.configMap[chain][property] ?? ethers.constants.AddressZero;
+        typeof this.configMap[chain][property] === 'string'
+          ? String(this.configMap[chain][property]) ??
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply typescript/sdk/src/router/HyperlaneRouterChecker.ts:45]: Fixed!

File: typescript/sdk/src/middleware/query/InterchainQueryDeployer.ts [hunk 5]
```diff
@@ -40,7 +40,9 @@ export class InterchainQueryDeployer extends ProxiedRouterDeployer<
     return [
       config.mailbox,
       config.interchainGasPaymaster,
-      config.interchainSecurityModule ?? ethers.constants.AddressZero,
+      typeof config.interchainSecurityModule === 'string'
+        ? config.interchainSecurityModule
+        : ethers.constants.AddressZero,
```
Comment (yorhodes) [inline typescript/sdk/src/middleware/query/InterchainQueryDeployer.ts:45]: ditto

File: typescript/sdk/src/providers/MultiProvider.ts [hunk 6]
```diff
@@ -230,7 +230,7 @@ export class MultiProvider {
 
     if (TestChains.includes(name as CoreChainName)) {
       this.providers[name] = new providers.JsonRpcProvider(
-        'http://localhost:8545',
+        'http://127.0.0.1:8545',
```
Comment (yorhodes) [inline typescript/sdk/src/providers/MultiProvider.ts:233]: not sure why you needed to change this everywhere

File: typescript/sdk/src/router/HyperlaneRouterDeployer.ts [hunk 7]
```diff
@@ -38,9 +41,14 @@ export abstract class HyperlaneRouterDeployer<
         signerOrProvider,
       );
       const localIsm =
-        config.interchainSecurityModule ?? (await localMailbox.defaultIsm());
+        !config.interchainSecurityModule ||
+        config.interchainSecurityModule === ethers.constants.AddressZero
+          ? await localMailbox.defaultIsm()
+          : config.interchainSecurityModule;
```
Comment (yorhodes) [inline typescript/sdk/src/router/HyperlaneRouterDeployer.ts:47]: this is hard to read

**Review by yorhodes — CHANGES_REQUESTED** [review]

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 8]
```diff
@@ -324,6 +380,16 @@ export async function moduleMatchesConfig(
...
+    case ModuleType.MESSAGE_ID_MULTISIG: {
+      // A MessageIdMultisigIsm matches if validators and threshold match the config
+      const expectedAddress =
+        await contracts.messageIdMultisigIsmFactory.getAddress(
+          config.validators.sort(),
+          config.threshold,
+        );
+      matches = utils.eqAddress(expectedAddress, module.address);
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply typescript/sdk/src/ism/HyperlaneIsmFactory.ts:392]: what can be deduped?

File: typescript/sdk/src/middleware/query/InterchainQueryDeployer.ts (hunk 5 above)
Comment (aroralanuk) [reply typescript/sdk/src/middleware/query/InterchainQueryDeployer.ts:45]: same as last comment.

File: typescript/sdk/src/providers/MultiProvider.ts (hunk 6 above)
Comment (aroralanuk) [reply typescript/sdk/src/providers/MultiProvider.ts:233]: wasn't working with anvil or hardhat, with error `ECONNREFUSED` from the rpc.

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 9]
```diff
@@ -234,65 +257,99 @@ export async function moduleCanCertainlyVerify(
...
+      }
+    } catch (e) {
+      logging.warn(`Error checking module ${destModule}: ${e}`);
+      return false;
+    }
+  } else {
+    // destModule is an IsmConfig
+    let verified = 0; // can't declare this in case statement
+    // checksConfig checks for origin to origin - which is undefined
+    if (destModule) {
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply typescript/sdk/src/ism/HyperlaneIsmFactory.ts:323]: fixed

File: typescript/infra/config/environments/testnet3/aggregationIsm.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        return 'MESSAGE_ID_MULTISIG';
+      default:
+        return value;
+    }
+  }
+  return value;
+};
+
+export const printIsmConfig = (ism: IsmConfig): string => {
+  return `IsmConfig: ${JSON.stringify(ism, replacerEnum, 2)}`;
```
Comment (yorhodes) [inline typescript/infra/config/environments/testnet3/aggregationIsm.ts:91]:
```suggestion
  return JSON.stringify(ism, replacerEnum, 2);
```

File: typescript/sdk/src/router/HyperlaneRouterChecker.ts [hunk 11]
```diff
@@ -39,8 +39,11 @@ export class HyperlaneRouterChecker<
       violationType: ConnectionClientViolationType,
     ) => {
       const actual = await router[property]();
+      // TODO: check for IsmConfig
       const expected =
-        this.configMap[chain][property] ?? ethers.constants.AddressZero;
+        typeof this.configMap[chain][property] === 'string'
+          ? `${this.configMap[chain][property]}` ?? ethers.constants.AddressZero
+          : ethers.constants.AddressZero;
```
Comment (yorhodes) [inline typescript/sdk/src/router/HyperlaneRouterChecker.ts:46]:
```suggestion
        typeof this.configMap[chain][property] === 'string'
          ? this.configMap[chain][property]
          : ethers.constants.AddressZero;
```

File: typescript/sdk/src/router/HyperlaneRouterChecker.ts [hunk 12]
```diff
@@ -39,8 +39,11 @@ export class HyperlaneRouterChecker<
       violationType: ConnectionClientViolationType,
     ) => {
       const actual = await router[property]();
+      // TODO: check for IsmConfig
```
Comment (yorhodes) [inline typescript/sdk/src/router/HyperlaneRouterChecker.ts:42]: throw if ism is object (unimplemented)

File: typescript/sdk/src/router/types.ts [hunk 13]
```diff
@@ -32,7 +33,7 @@ export type ProxiedFactories = HyperlaneFactories & {
 export type ConnectionClientConfig = {
   mailbox: types.Address;
   interchainGasPaymaster: types.Address;
-  interchainSecurityModule?: types.Address;
+  interchainSecurityModule?: types.Address | IsmConfig;
```
Comment (yorhodes) [inline typescript/sdk/src/router/types.ts:36]: nice

**Review by yorhodes — APPROVED** [review]

**Pull request merged**